### PR TITLE
WWST-959 add orvibo gas detector DTH

### PR DESCRIPTION
**Pull request by jinkangzhangS**
Plz help review the orvibo gas detector DTH

### CODE REVIEW

File: devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ *  Author : jinkang zhang / jk0218.zhang@samsung.com
+ *  Date : 2018-07-04
+ */
+import physicalgraph.zigbee.clusters.iaszone.ZoneStatus
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "Orvibo Gas Detector", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn: "SmartThings", vid: "generic-smoke") {
```
Comment (greens) [inline devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy]: this name field needs to match the file and directory name.

File: devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		standardTile("main", "device.smoke", width: 2, height: 2) {
+			state("clear", label:"Clear", icon:"st.alarm.smoke.clear", backgroundColor:"#ffffff")
+			state("detected", label:"Smoke!", icon:"st.alarm.smoke.smoke", backgroundColor:"#e86d13")
+		}
+
+		standardTile("alarm", "device.alarm", width: 2, height: 2) {
+                state "off", label:'off', action:'alarm.strobe', icon:"st.alarm.alarm.alarm", backgroundColor:"#ffffff"
+                state "strobe", label:'strobe!', action:'alarm.off', icon:"st.alarm.alarm.alarm", backgroundColor:"#e86d13"
+                state "siren", label:'siren!', action:'alarm.off', icon:"st.alarm.alarm.alarm", backgroundColor:"#e86d13"
+                state "both", label:'alarm!', action:'alarm.off', icon:"st.alarm.alarm.alarm", backgroundColor:"#e86d13"
```
Comment (greens) [inline devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy]: missing a `}`

**Review by greens — CHANGES_REQUESTED** [review]
Please make the whitespace consistent and please add an initial state poll in the `installed` or `configure` methods.

File: devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy (hunk 1 above)
Comment (jinkangzhangS) [reply devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy]: ok,  i will rename the direcory name .

**Review by jinkangzhangS — COMMENTED** [review]
Hi ,Mr greens , how to use the "initial state poll" , is there any document about it ?

**Comment by greens** [comment]
What I meant by "poll initial state" is to query the device to have it report the values of its various attributes when it is first added to the system. For this device that would mean querying the zigbee attributes responsible for the smoke detection and alarm state.

Also does this device use a battery? If so, it should be incorporated into the DTH.

**Comment by jinkangzhangS** [comment]
Orvibo Gas detecter did not support command configure report !

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ *  Author : jinkang zhang / jk0218.zhang@samsung.com
+ *  Date : 2018-07-04
+ */
+import physicalgraph.zigbee.clusters.iaszone.ZoneStatus
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "Orvibo Gas Detector", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn: "SmartThings", vid: "generic-smoke") {
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy:22]: What does Gas detector mean. Does it detect smoke or Carbon Monoxide or both. Maybe new capability need to be created to that device. Please give more information about that device

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		capability "Sensor"
+		capability "Refresh"
+		fingerprint profileId: "0104", deviceId: "0402", inClusters: "0000,0003,0500,0009", outClusters: "0019",manufacturer: "Heiman",model:"d0e857bfd54f4a12816295db3945a421"
+	}
+	tiles {
+		standardTile("main", "device.smoke", width: 2, height: 2) {
+			state("clear", label:"Clear", icon:"st.alarm.smoke.clear", backgroundColor:"#ffffff")
+			state("detected", label:"Smoke!", icon:"st.alarm.smoke.smoke", backgroundColor:"#e86d13")
+		}
+		standardTile("alarm", "device.alarm", width: 2, height: 2) {
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: Is this device also siren or Strobe. You added Alarm capability but is it possible to start siren for that device by other device like motion sensor?

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ *  Date : 2018-07-04
+ */
+import physicalgraph.zigbee.clusters.iaszone.ZoneStatus
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "Orvibo Gas Detector", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn: "SmartThings", vid: "generic-smoke") {
+		capability "Smoke Detector"
+		capability "Configuration"
+		capability "Alarm"
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: Is this device also siren or Strobe. You added Alarm capability but is it possible to start siren for that device by other device like motion sensor?

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+}
+private Map getDetectedResult(value) {
+	String descriptionText = value == 'detected' ? "${device.displayName} detected smoke" : "${device.displayName} smoke clear"
+	if(value == detected){
+		strobe()
+	}else{
+		off()
+	}
+	return [
+			name			: 'smoke',
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: formatting

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+private Map parseIasMessage(String description) {
+	ZoneStatus zs = zigbee.parseZoneStatus(description)
+	translateZoneStatus(zs)
+}
+private Map translateZoneStatus(ZoneStatus zs) {
+	// Some sensor models that use this DTH use alarm1 and some use alarm2 to signify motion
+	return (zs.isAlarm1Set() || zs.isAlarm2Set()) ? getDetectedResult('detected') : getDetectedResult('clear')
+}
+private Map getDetectedResult(value) {
+	String descriptionText = value == 'detected' ? "${device.displayName} detected smoke" : "${device.displayName} smoke clear"
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: refactor this using value like:
"${device.displayName} smoke ${value}"

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	if(value == detected){
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: remove this if alarm capability is not necessary

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+		off()
+	}
+	return [
+			name			: 'smoke',
+			value			: value,
+			descriptionText : descriptionText,
+			translatable	: true
+	]
+}
+def strobe() {
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: remove those four method i alarm capability is not necessary

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ * */
+def ping() {
+	log.debug "ping "
+	return zigbee.readAttribute(zigbee.IAS_ZONE_CLUSTER, zigbee.ATTRIBUTE_IAS_ZONE_STATUS)
+}
+def configure() {
+	log.debug "configure"
+	sendEvent(name: "checkInterval", value:40 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: Due to possibility of network lag add some additional timeout to this like 2 mins

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+}
+private Map parseIasMessage(String description) {
+	ZoneStatus zs = zigbee.parseZoneStatus(description)
+	translateZoneStatus(zs)
+}
+private Map translateZoneStatus(ZoneStatus zs) {
+	// Some sensor models that use this DTH use alarm1 and some use alarm2 to signify motion
+	return (zs.isAlarm1Set() || zs.isAlarm2Set()) ? getDetectedResult('detected') : getDetectedResult('clear')
+}
+private Map getDetectedResult(value) {
```
Comment (MAblewicz) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: You are creating new method (translateZoneStatus) just to call return without any additional logic. Just call return here.
Or even better. Change getMotionResult(value) method. Check if value is true or false and create events based on that. Remeber to change stopMotion().

**Review by MAblewicz — CHANGES_REQUESTED** [review]

File: devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy (hunk 1 above)
Comment (greens) [reply devicetypes/smartthings/Orvibo-Gas-detect.src/Orvibo-Gas-detect.groovy]: You'll need a new commit that has deleted this file.

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy (hunk 3 above)
Comment (jinkangzhangS) [reply devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy:22]: Gas detector means flammable gas detector.

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy (hunk 4 above)
Comment (jinkangzhangS) [reply devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: It is same with smoke detector , when detect Gas, it will alarm .

**Review by greens — APPROVED** [review]

**Review by greens — CHANGES_REQUESTED** [review]
Can you actually turn the siren on and off? The `siren()` and `off()` method doesn't send anything to the device.

**Review by greens — DISMISSED** [review]

**Comment by greens** [comment]
This looks fine to me. @MAblewiczS ?

File: devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy [hunk 12]
```diff
@@ -94,6 +94,7 @@ def ping() {
 }
 def configure() {
 	log.debug "configure"
-	//The gas heart beat is 30 mintues, so we set double time for one package loss.
-	sendEvent(name: "checkInterval", value: 30 * 2 * 60 + 2*60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
+	//The gas sensor has no battery , it can response the ping immediately
+	//so i set the checkTinterval to 5 minutes .
+	sendEvent(name: "checkInterval", value: 5*60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
```
Comment (greens) [inline devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: Health Check is not meant to be used as a polling service. It's meant to verify that devices have not gone offline. Additionally, what user experience is improved by polling the device every five minutes? The device should be notifying us of its changes. If it's not, there should be configuration options to adjust the reporting interval.

Comment (jinkangzhangS) [reply devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: The tester said the 30 minutes is too long to be offline when the gas is remove.
The test case need the gas sensor to be offline in five minutes when it is removed from the hub.

Comment (greens) [reply devicetypes/smartthings/orvibo-gas-detector.src/Orvibo-Gas-detector.groovy]: That requirement should be adjusted. Most of our sensors have a 32 minute health check and should be marked offline after 33 minutes of being unplugged or taken out of range.

**Review by MAblewicz — APPROVED** [review]
With now correct interval 32min it looks ok.  I will approve this PR

**Comment by workingmonk** [comment]
@MAblewiczS This should have been squashed and merged or ask the developer to rebase. We just added 10 commits for a single change

**Comment by MAblewicz** [comment]
@workingmonk thanks for information. It was my first merge here. Now i know that squash is required. Next time i will squash and merge requests.

**Comment by workingmonk** [comment]
cool. "Squash and merge" is part of the drop down while merging.

**Pull request merged**